server/src/handlers/prom.rs

```diff
@@ -0,0 +1,427 @@
+    MissingName { backtrace: Backtrace },
+
+    #[snafu(display("Invalid matcher type, value:{}.\nBacktrace:\n{}", value, backtrace))]
+    InvalidMatcherType { value: i32, backtrace: Backtrace },
+
+    #[snafu(display("Read response must be Rows.\nBacktrace:\n{}", backtrace))]
+    ResponseMustRows { backtrace: Backtrace },
+
+    #[snafu(display("TSID column is missing in query response.\nBacktrace:\n{}", backtrace))]
+    MissingTSID { backtrace: Backtrace },
+
+    #[snafu(display(
+        "Timestamp column is missing in query response.\nBacktrace:\n{}",
+        backtrace
+    ))]
+    MissingTimestamp { backtrace: Backtrace },
+
+    #[snafu(display(
+        "Value column is missing in query response.\nBacktrace:\n{}",
+        backtrace
+    ))]
+    MissingValue { backtrace: Backtrace },
+
+    #[snafu(display("Handle sql failed, err:{}.", source))]
+    SqlHandle {
+        source: Box<crate::handlers::error::Error>,
+    },
+
+    #[snafu(display("Tsid must be u64, current:{}.\nBacktrace:\n{}", kind, backtrace))]
+    TsidMustU64 {
+        kind: DatumKind,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Timestamp wrong type, current:{}.\nBacktrace:\n{}", kind, backtrace))]
+    MustTimestamp {
+        kind: DatumKind,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "Value must be f64 compatible type, current:{}.\nBacktrace:\n{}",
+        kind,
+        backtrace
+    ))]
+    F64Castable {
+        kind: DatumKind,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "Tag must be string type, current:{}.\nBacktrace:\n{}",
+        kind,
+        backtrace
+    ))]
+    TagMustString {
+        kind: DatumKind,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Failed to write via gRPC, source:{}.", source))]
+    GRPCWriteError {
+        source: crate::grpc::storage_service::error::Error,
+    },
+
+    #[snafu(display("Failed to get schema, source:{}.", source))]
+    SchemaError {
+        source: crate::schema_config_provider::Error,
+    },
+}
+
+define_result!(Error);
+
+impl reject::Reject for Error {}
+
+const NAME_LABEL: &str = "__name__";
+const VALUE_COLUMN: &str = "value";
+
+pub struct CeresDBStorage<Q: QueryExecutor + 'static> {
+    instance: InstanceRef<Q>,
+    schema_config_provider: SchemaConfigProviderRef,
+}
+
+impl<Q: QueryExecutor + 'static> CeresDBStorage<Q> {
+    pub fn new(instance: InstanceRef<Q>, schema_config_provider: SchemaConfigProviderRef) -> Self {
+        Self {
+            instance,
+            schema_config_provider,
+        }
+    }
+}
+
+impl<Q: QueryExecutor + 'static> CeresDBStorage<Q> {
+    /// This function will separate measurement from labels, and sort labels by
+    /// name.
+    fn normalize_labels(labels: Vec<Label>) -> Result<(String, Vec<Label>)> {
+        let mut new_labels = Vec::with_capacity(labels.len());
+        let mut measurement = None;
```

Comment:
In order to extract the metric from the `labels`, `swap_move` may be more efficient than creating a new vector.

Reply:
👍
Fixed, this function looks more simple and elegant now.